### PR TITLE
Generic to dockerfile

### PR DESCRIPTION
**Pull request by 0xDC0DE**
See progress in ticket here: https://github.com/returntocorp/semgrep-rules/issues/1793

### CODE REVIEW

File: dockerfile/best-practice/missing-zypper-clean.yaml [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
   - id: missing-zypper-clean
     severity: WARNING
-    languages: [generic]
+    languages: [dockerfile]
     patterns:
-      - pattern: zypper $COMMAND
-      - pattern-not-inside: RUN ... && zypper clean
-      - pattern-not-inside: RUN ... && \ zypper clean
+      - pattern: RUN ... zypper ...
+      - pattern-not: RUN ... zypper clean
```
Comment (kurt-r2c) [inline dockerfile/best-practice/missing-zypper-clean.yaml]: We removed enforcing the `&&` condition here - intentional?

Edit: If the user isn't using e.g. `||` over `&&` they could break their container build by having a failed `zypper` invoke followed by a successful `zypper clean` which would look "truthy" to output and cause the layer construction to "succeed".

File: dockerfile/best-practice/prefer-copy-over-add.yaml [hunk 2]
```diff
@@ -17,4 +17,4 @@ rules:
       include:
         - "*dockerfile*"
         - "*Dockerfile*"
-    pattern: ADD
+    pattern: ADD $FROM $TO
```
Comment (kurt-r2c) [inline dockerfile/best-practice/prefer-copy-over-add.yaml:20]: Can this be an ellipsis or do we need to capture these?

File: dockerfile/best-practice/remove-package-cache.yaml [hunk 3]
```diff
@@ -1,15 +1,13 @@
 rules:
   - id: remove-package-cache
     patterns:
-      - pattern: RUN apt-get update
-      - pattern-not-inside: RUN apt-get update ... && apt-get clean
-      - pattern-not-inside: RUN apt-get update ... &&\ apt-get clean
+      - pattern: RUN ... apt-get update ...
+      - pattern-not: RUN ... apt-get clean ...
```
Comment (kurt-r2c) [inline dockerfile/best-practice/remove-package-cache.yaml]: same comment as before re: `zypper` - maybe a separate rule for "incorrect control operators" which is out of scope for this PR.

**Review by kurt-r2c — COMMENTED** [review]
This looks OK. We should consider a followup rule for incorrect operator usage in the `RUN` directive. 

Also, it looks like the migration from `generic/dockerfile/**` to `dockerfile/**` is incomplete.  Did some of these rules not get moved on purpose or is this an oversight?

**Comment by kurt-r2c** [comment]
This PR is a WIP - see linked ticket from @pieterdc1

**Comment by kurt-r2c** [comment]
blocked by: 
https://github.com/returntocorp/semgrep/issues/4866
https://github.com/returntocorp/semgrep/issues/4881
https://github.com/returntocorp/semgrep/issues/4875

File: dockerfile/best-practice/remove-package-cache.yaml (hunk 3 above)
Comment (0xDC0DE) [reply dockerfile/best-practice/remove-package-cache.yaml]: We actually have a rule for that, set-pipefail.

**Comment by 0xDC0DE** [comment]
I would opt to merge the ported rules and create separate pull request for the additional rules when those are unblocked.

File: dockerfile/best-practice/remove-package-cache.yaml (hunk 3 above)
Comment (0xDC0DE) [reply dockerfile/best-practice/remove-package-cache.yaml]: Updated test cases to include `||` operator.

**Review by kurt-r2c — APPROVED** [review]
LGTM

**Comment by kurt-r2c** [comment]
@pieterdc1 can you resolve the merge conflicts in `avoid-latest-version` and `missing-pip-no-cache-dir`?

**Pull request merged**